### PR TITLE
Fix problems with EINTR in ExternalDataSource::getline 

### PR DESCRIPTION
**Pull request by bzeller**

### CODE REVIEW

File: zypp-core/zyppng/base/private/linuxhelpers_p.h [hunk 1]
```diff
@@ -21,7 +21,8 @@ namespace zyppng {
   auto eintrSafeCall ( Fun &&function, Args&&... args ) {
     int res;
     do {
-      res = std::forward<Fun>(function)( std::forward<Args>(args)... );
+      errno = 0;
+      res   = std::forward<Fun>(function)( std::forward<Args>(args)... );
```
Comment (mlandres) [inline zypp-core/zyppng/base/private/linuxhelpers_p.h:25]:
```suggestion
      res  = std::forward<Fun>(function)( std::forward<Args>(args)... );
```

**Review by mlandres — APPROVED** [review]
Just the double space...

**Comment by mlandres** [comment]
> Just the double space...

NO, aligned ;)

File: zypp-core/zyppng/base/private/linuxhelpers_p.h (hunk 1 above)
Comment (mlandres) [reply zypp-core/zyppng/base/private/linuxhelpers_p.h:25]: Now I see: Alligned with errno =

File: zypp-core/zyppng/base/private/linuxhelpers_p.h [hunk 2]
```diff
@@ -17,15 +17,40 @@ namespace zyppng {
...
+   */
+  template<typename Fun, typename RestartCb, typename... Args >
+  auto eintrSafeCallEx ( Fun &&function, RestartCb &&restartCb, Args&&... args ) {
     int res;
-    do {
-      res = std::forward<Fun>(function)( std::forward<Args>(args)... );
-    } while ( res == -1 && errno == EINTR );
+    while ( true ) {
+      errno = 0;
+      res = function( args... );
```
Comment (mlandres) [inline zypp-core/zyppng/base/private/linuxhelpers_p.h:39]:
```suggestion
      res = std::forward<Fun>(function)( std::forward<Args>(args)... );
```

File: zypp-core/zyppng/base/private/linuxhelpers_p.h [hunk 3]
```diff
@@ -17,15 +17,40 @@ namespace zyppng {
...
+      if ( res == -1 && errno == EINTR ) {
+        std::invoke( restartCb );
```
Comment (mlandres) [inline zypp-core/zyppng/base/private/linuxhelpers_p.h:41]:
```suggestion
        std::invoke( std::forward<RestartCb>(restartCb) );
```

**Review by mlandres — CHANGES_REQUESTED** [review]

File: zypp-core/zyppng/base/private/linuxhelpers_p.h (hunk 2 above)
Comment (bzeller) [reply zypp-core/zyppng/base/private/linuxhelpers_p.h:39]: What if we get a rvalue reference ? The calling function would be free to move the object away , so in a 2nd iteration we would have undefined behavior?

File: zypp-core/zyppng/base/private/linuxhelpers_p.h (hunk 3 above)
Comment (bzeller) [reply zypp-core/zyppng/base/private/linuxhelpers_p.h:41]: Same as above

**Review by mlandres — APPROVED** [review]

**Pull request merged**